source-mongodb/pull.go

```diff
@@ -132,141 +109,213 @@ func (d *driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) err
+	// have readAnyDatabase access on the instance. If this fails for permission
+	// issues, we then attempt to do a database-level change stream. In the case
+	// of the database-level change stream we expect only a single database to be
+	// provided and do not support multiple databases in this mode.
+	cursor, err := client.Watch(ctx, mongo.Pipeline{eventFilter}, opts)
 	if err != nil {
 		// This error means events from the starting point are no longer available,
 		// which means we have hit a gap in between last run of the connector and
-		// this one. We re-set the checkpoint and error out so the next run of the
-		// connector will run a backfill.
-		//
-		// There is potential for optimisation here by only resetting the checkpoint
-		// of the collection which can't resume instead of resetting the whole
-		// connector's checkpoint and backfilling everything
+		// this one.
 		if e, ok := err.(mongo.ServerError); ok {
 			if e.HasErrorMessage(resumePointGoneErrorMessage) || e.HasErrorMessage(resumeTokenNotFoundErrorMessage) {
-				return fmt.Errorf("change stream on collection %s cannot resume capture, this is usually due to a small oplog storage available. Please resize your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE. After resizing your uplog, you can remove the binding for this collection add it back to trigger a backfill: %w", res.Collection, e)
+				cp, err := json.Marshal(captureState{})
+				if err != nil {
+					return fmt.Errorf("marshalling checkpoint for removed bindings: %w", err)
+				} else if err = c.Output.Checkpoint(cp, false); err != nil {
+					return fmt.Errorf("updating checkpoint for removed bindings: %w", err)
+				}
+				return fmt.Errorf("change stream cannot resume capture, this is usually due to insufficient oplog storage. Please consider resizing your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE. The connector will now attempt to backfill all bindings again: %w", e)
+			} else if e.HasErrorMessage(unauthorizedMessage) {
+				log.WithField("err", e).Warn("could not start instance-level change stream, trying database-level change stream.")
+				// Check that there is only one database, if so, we can start a change
+				// stream on this database. If there is more than one database
+				// configured, we error out as we do not support multiple change streams
+				// on different databases. An instance-level change stream must be used
+				// for that scenario
+				var dbName = resources[0].Database
+				for _, res := range resources[1:] {
+					if res.Database != dbName {
+						return fmt.Errorf("We are unable to establish an instance-level change stream on your database due to an error. Please consider granting permission to read all databases to the user: %w", err)
+					}
+				}
+
+				var db = client.Database(dbName)
+				// TODO: what happens if a user switches from readAnyDatabase to
```

Comment:
This does seem like an important consideration. I don't know how resume tokens work for a "global" change stream vs a per-database one. Would switching permissions from read@<db> to readAnyDatabase break the capture? Presumably _increasing_ the connector's permissions shouldn't break things, but if the resume tokens aren't compatible then it would, which wouldn't be good. If that does break things, I think ideally we would store the type of permission being used in the state checkpoint, only emitting it with the very first checkpoint, and then always continue to try to use that mechanism.

Reply:
Had a chat with Johnny on this as well, yes I think using the state to keep track of the mode of change stream is important: we will still break, but we will break understanding why and can help the user. The idea is to keep track of the mode, and if there is a change in the mode, we automatically do a backfill and switch to the new mode.

Reply:
Actually I did a test of this scenario and turns out the resume token is not invalid across these two change streams. In this case I think it is okay to continue using the resume token of the other change stream, since that means we already have covered all the events relevant to the actively captured databases / collections. If a new database / collection is added, we will inevitably backfill it, so it should also be okay in that case.